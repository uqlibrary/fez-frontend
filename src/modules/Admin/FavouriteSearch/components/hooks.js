import { useCallback, useState } from 'react';
import { GridRowModes } from '@mui/x-data-grid';

export const useDecoratedDataGrid = (list, handleRowUpdate, handleRowDelete) => {
    const [rows, setRows] = useState(list);
    const [loading, setLoading] = useState(false);
    const [rowModesModel, setRowModesModel] = useState({});
    const [deleteRowId, setDeleteRowId] = useState(null);

    const handleUpdateRow = useCallback(
        async (newData, oldData) => {
            setLoading(true);
            const res = await handleRowUpdate(newData, oldData);
            setLoading(false);
            return res;
        },
        [handleRowUpdate],
    );

    const handleDeleteRow = useCallback(
        (id, rows) => async () => {
            setLoading(true);
            const rowToDelete = rows.find(row => row.fvs_id === id);
            const newRows = await handleRowDelete(rows, rowToDelete);
            setRows(newRows);
            setDeleteRowId(null);
            setLoading(false);
        },
        [handleRowDelete],
    );

    const handleRowModesModelChange = useCallback(newRowModesModel => {
        setRowModesModel(newRowModesModel);
    }, []);

    const handleEditClick = useCallback(
        id => () => {
            setRowModesModel({
                [id]: { mode: GridRowModes.Edit },
            });
        },
        [],
    );

    const handleSaveClick = useCallback(
        id => () => {
            setRowModesModel({ ...rowModesModel, [id]: { mode: GridRowModes.View } });
        },
        [rowModesModel],
    );

    const handleDeleteClick = useCallback(
        id => () => {
            setDeleteRowId(id);
        },
        [],
    );

    const handleCancelClick = useCallback(
        id => () => {
            setRowModesModel({
                ...rowModesModel,
                [id]: { mode: GridRowModes.View, ignoreModifications: true },
            });

            setDeleteRowId(null);
        },
        [rowModesModel],
    );

    return {
        loading,
        deleteRowId,
        rows,
        rowModesModel,
        handleUpdateRow,
        handleDeleteRow,
        handleRowModesModelChange,
        handleEditClick,
        handleSaveClick,
        handleDeleteClick,
        handleCancelClick,
    };
};
