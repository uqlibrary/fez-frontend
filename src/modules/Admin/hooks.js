import { useRef } from 'react';
import { useSelector } from 'react-redux';
import { getInitialFormValues } from './helpers';
import {
    RECORD_TYPE_RECORD,
    PUBLICATION_TYPE_THESIS,
    PUBLICATION_TYPE_CONFERENCE_PAPER,
    PUBLICATION_TYPE_JOURNAL_ARTICLE,
} from 'config/general';
import { useWatch } from 'react-hook-form';
import { unionBy } from 'lodash';

export const useInitialFormValues = (recordToView, createMode) => {
    const initialFormValues = {
        initialValues: {
            bibliographicSection: {
                languages: ['eng'],
            },
        },
    };

    if (createMode) return initialFormValues;
    const recordType = ((recordToView || {}).rek_object_type_lookup || '').toLowerCase();
    return (!!recordToView && recordToView.rek_pid && getInitialFormValues(recordToView, recordType)) || {};
};
export const useRecord = (displayType, subType, createMode) => {
    const { authorDetails, author } = useSelector(state => state.get('accountReducer'));
    const {
        recordToView,
        isRecordLocked,
        loadingRecordToView,
        isDeleted,
        isJobCreated,
        recordToViewError,
        error,
    } = useSelector(state => state.get('viewRecordReducer'));

    const initialFormValues = useInitialFormValues(recordToView, createMode);

    return {
        loadingRecordToView,
        authorDetails: authorDetails ?? /* istanbul ignore next */ null,
        author,
        isDeleted,
        isJobCreated,
        recordToView,
        recordToViewError,
        ...initialFormValues,
        locked: isRecordLocked || false,
        error,
    };
};

export const useRecordToView = (recordToView, createMode, methods) => {
    const { newRecord } = useSelector(state => (createMode ? state.get('createAdminRecordReducer') : {}));

    if (!createMode) return recordToView;

    const displayType = methods.getValues('rek_display_type');
    const selectedSubType = methods.getValues('adminSection.rek_subtype');

    return {
        rek_pid: newRecord?.rek_pid ?? null,
        rek_display_type: displayType,
        rek_subtype: selectedSubType,
        rek_object_type_lookup: RECORD_TYPE_RECORD,
    };
};

export const useFormOnChangeHook = form => {
    const prevBibliographicSectionFezMatchedJournals = useRef('');
    const [
        rekDisplayType,
        adminSectionRekSubtype,
        bibliographicSectionRekGenreType,
        bibliographicSectionFezMatchedJournals,
    ] = useWatch({
        control: form.control,
        name: [
            'rek_display_type',
            'adminSection.rek_subtype',
            'bibliographicSection.rek_genre_type',
            'bibliographicSection.fez_matched_journals',
        ],
    });
    if (rekDisplayType === PUBLICATION_TYPE_THESIS && !!adminSectionRekSubtype && !!!bibliographicSectionRekGenreType) {
        console.log('updating bibliographicSection.rek_genre_type', adminSectionRekSubtype);
        form.setValue('bibliographicSection.rek_genre_type', adminSectionRekSubtype);
    }

    const { isTouched } = form.getFieldState('bibliographicSection.fez_matched_journals');

    if (
        bibliographicSectionFezMatchedJournals &&
        bibliographicSectionFezMatchedJournals.id !== prevBibliographicSectionFezMatchedJournals.current &&
        isTouched === false &&
        [PUBLICATION_TYPE_CONFERENCE_PAPER, PUBLICATION_TYPE_JOURNAL_ARTICLE].includes(rekDisplayType)
    ) {
        prevBibliographicSectionFezMatchedJournals.current = bibliographicSectionFezMatchedJournals.id;
        const issns =
            bibliographicSectionFezMatchedJournals?.fez_journal_issn?.map(issn => ({
                rek_value: {
                    key: issn.jnl_issn,
                    value: {
                        sherpaRomeo: { link: false },
                        ulrichs: { link: false, linkText: '' },
                    },
                },
            })) || [];
        form.setValue(
            'bibliographicSection.fez_record_search_key_journal_name.rek_journal_name',
            bibliographicSectionFezMatchedJournals.value,
            { shouldValidate: true, shouldDirty: true },
        );
        const bibliographicSectionIssns = form.getValues('bibliographicSection.issns') || [];
        // create new array, filter out any dupes
        const updatedBibliographicSectionIssns = unionBy(bibliographicSectionIssns, issns, 'rek_value.key');
        form.setValue('bibliographicSection.issns', updatedBibliographicSectionIssns);
    }
};
