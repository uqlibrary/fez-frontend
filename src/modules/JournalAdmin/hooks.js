import { useMemo } from 'react';
import { useSelector } from 'react-redux';
import { getInitialFormValues } from './helpers';

export const useJournal = () => {
    const {
        journalToView,
        isJournalLocked,
        loadingJournalToView,
        journalToViewError,
        journalLoadingError,
        error,
    } = useSelector(state => state.get('viewJournalReducer'));

    const { authorDetails, author } = useSelector(state => state.get('accountReducer'));
    const values = useMemo(() => {
        const initialFormValues =
            (!!journalToView && journalToView.jnl_jid && getInitialFormValues(journalToView)) || {};
        return {
            journalToViewLoading: loadingJournalToView,
            authorDetails: authorDetails ?? /* istanbul ignore next */ null,
            author,
            journalToView,
            journalToViewError,
            journalLoadingError,
            ...initialFormValues,
            locked: isJournalLocked,
            error,
        };
    }, [
        author,
        authorDetails,
        isJournalLocked,
        journalLoadingError,
        journalToView,
        journalToViewError,
        loadingJournalToView,
        error,
    ]);
    return { ...values };
};
