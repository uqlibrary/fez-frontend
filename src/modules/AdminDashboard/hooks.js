import React from 'react';

import { useDispatch } from 'react-redux';
import { useConfirmationState } from 'hooks';
import { isEmptyStr } from './utils';

export const useSystemAlertDrawer = data => {
    const [_open, _setOpen] = React.useState(false);
    const [_row, _setRow] = React.useState({});

    React.useEffect(() => {
        if (_open) {
            _setRow(data?.find(item => item.sat_id === _row.sat_id) || /* istanbul ignore next */ {});
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [data]);

    const openDrawer = row => {
        _setRow(row);
        _setOpen(true);
    };
    const closeDrawer = () => {
        _setRow({});
        _setOpen(false);
    };
    return {
        open: _open,
        row: _row,
        openDrawer,
        closeDrawer,
    };
};

export const useAlertStatus = ({ message, hideAction }) => {
    const dispatch = useDispatch();
    const [alertIsVisible, _showAlert, _hideAlert] = useConfirmationState();

    const [_message, setMessage] = React.useState(message);

    const showAlert = message => {
        setMessage(message);
        _showAlert();
    };

    const hideAlert = () => {
        hideAction && dispatch(hideAction());
        _hideAlert();
    };

    React.useEffect(() => {
        if (!alertIsVisible) {
            if (!isEmptyStr(message)) _showAlert();
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [alertIsVisible, message]);

    return [alertIsVisible, hideAlert, showAlert, _message];
};
