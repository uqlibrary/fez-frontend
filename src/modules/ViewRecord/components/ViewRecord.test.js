import React from 'react';
import ViewRecord from './ViewRecord';
import Immutable from 'immutable';
import { fireEvent, render, WithReduxStore, WithRouter, createMatchMedia } from 'test-utils';
import * as ViewRecordActions from 'actions/viewRecord';
import { userIsAdmin, userIsAuthor } from 'hooks';
import { ntro } from 'mock/data/testing/records';
import { default as record } from 'mock/data/records/record';
import { recordWithNoAffiliationIssues } from 'mock/data/records';
import { accounts, currentAuthor } from 'mock/data/account';
import { useParams } from 'react-router-dom';
import { recordVersionLegacy } from 'mock/data';
import locale from 'locale/pages';
import { notFound } from 'config/routes';
import globalLocale from 'locale/global';
import { default as recordWithNotes } from 'mock/data/records/recordWithNotes';
import { default as recordWithAuthorAffiliates } from 'mock/data/records/recordWithAuthorAffiliates';
import { NTRO_SUBTYPE_RREB_PUBLIC_SECTOR, PUBLICATION_TYPE_DATA_COLLECTION } from 'config/general';

jest.mock('../../../hooks', () => ({
    userIsAdmin: jest.fn(() => ({})),
    isIntWithin: jest.requireActual('../../../hooks').isIntWithin,
    belongsToAuthor: jest.requireActual('../../../hooks').belongsToAuthor,
}));

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: jest.fn(() => ({ pid: 'UQ:123456' })),
    useNavigate: jest.fn(() => jest.fn()),
}));

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useDispatch: () => mockDispatch,
}));

const setup = (state = {}, renderer = render) => {
    const { accountReducer, viewRecordReducer } = state;
    const initState = {
        accountReducer: {
            account: accounts.uqresearcher,
            author: null,
            ...accountReducer,
        },
        viewRecordReducer: {
            isDeleted: false,
            isDeletedVersion: false,
            loadingRecordToView: false,
            recordToViewError: null,
            recordToView: null,
            ...viewRecordReducer,
        },
    };
    return renderer(
        <WithReduxStore initialState={Immutable.Map(initState)}>
            <WithRouter>
                <ViewRecord />
            </WithRouter>
        </WithReduxStore>,
    );
};

describe('ViewRecord', () => {
    beforeAll(() => {
        window.matchMedia = createMatchMedia(window.innerWidth);
    });

    beforeEach(() => {
        userIsAdmin.mockImplementation(() => false);
        userIsAuthor && userIsAuthor.mockImplementation(() => true);
    });

    afterEach(() => {
        userIsAdmin.mockReset();
        userIsAuthor && userIsAuthor.mockReset();
    });

    it('should render default empty view', () => {
        const { asFragment } = setup({});
        expect(asFragment()).toMatchInlineSnapshot(`
            <DocumentFragment>
              <div
                class="empty"
              />
            </DocumentFragment>
        `);
    });

    it('should not render components for empty record', () => {
        // Checked OK
        const { asFragment } = setup({
            viewRecordReducer: { recordToView: {} },
        });
        expect(asFragment()).toMatchInlineSnapshot(`
            <DocumentFragment>
              <div
                class="empty"
              />
            </DocumentFragment>
        `);
    });

    it('should render default view with admin menu', () => {
        // Checked OK
        userIsAdmin.mockImplementationOnce(() => true);
        const { getByTestId } = setup({
            viewRecordReducer: { recordToView: record },
        });

        expect(getByTestId('admin-actions-button')).toBeInTheDocument();
    });

    it('should render default view with admin menu when AA issues and internal notes exist', () => {
        // Checked OK
        const affiliationIssues = {
            fez_author_affiliation: [
                {
                    af_id: 478894,
                    af_pid: 'UQ:871c1f8',
                    af_author_id: 7624000,
                    af_percent_affiliation: 900,
                    af_org_id: 881,
                    af_status: 1,
                    fez_author: {
                        aut_id: 88844,
                        aut_display_name: 'one Wrong',
                    },
                    fez_org_structure: [
                        {
                            org_id: 881,
                            org_title: 'School of Chemistry and Molecular Biosciences',
                        },
                    ],
                },
                {
                    af_id: 478894,
                    af_pid: 'UQ:871c1f8',
                    af_author_id: 7624001,
                    af_percent_affiliation: 100000,
                    af_org_id: 881,
                    af_status: 1,
                    fez_author: {
                        aut_id: 88844,
                        aut_display_name: 'two OK',
                    },
                    fez_org_structure: [
                        {
                            org_id: 881,
                            org_title: 'School of Chemistry and Molecular Biosciences',
                        },
                    ],
                },
            ],
            fez_internal_notes: { ain_detail: 'test' },
        };
        const recordwithIssues = { ...recordWithNoAffiliationIssues, ...affiliationIssues };
        userIsAdmin.mockImplementationOnce(() => true);
        const { getByTestId } = setup({
            viewRecordReducer: { recordToView: recordwithIssues },
        });

        expect(getByTestId('admin-actions-button')).toBeInTheDocument();
    });

    it('should render default view with admin menu when no AA issues exist', () => {
        // Checked OK
        userIsAdmin.mockImplementationOnce(() => true);
        const { getByTestId } = setup({
            viewRecordReducer: { recordToView: recordWithNoAffiliationIssues },
        });
        expect(getByTestId('admin-actions-button')).toBeInTheDocument();
    });
    it('should render internal notes view with admin menu when no AA issues exist', () => {
        // Checked OK
        userIsAdmin.mockImplementationOnce(() => true);
        const { getByTestId } = setup({
            viewRecordReducer: {
                recordToView: { ...recordWithNoAffiliationIssues, fez_internal_notes: { ain_detail: 'test' } },
            },
        });
        expect(getByTestId('admin-actions-button')).toBeInTheDocument();
    });

    it('should render version', () => {
        // Checked OK
        const txt = locale.pages.viewRecord.version;
        const pid = 'UQ:1';
        const loadRecordToViewFn = jest.spyOn(ViewRecordActions, 'loadRecordVersionToView');
        useParams.mockImplementationOnce(() => ({ pid, version: recordVersionLegacy.rek_version }));
        const { getByTestId } = setup({ viewRecordReducer: { recordToView: recordVersionLegacy } });
        expect(loadRecordToViewFn).toHaveBeenCalledWith(pid, recordVersionLegacy.rek_version);
        expect(getByTestId(txt.alert.version.alertId)).toBeInTheDocument();
        expect(getByTestId(txt.alert.warning.alertId)).toBeInTheDocument();
    });

    it('should render deleted version', () => {
        // Checked OK
        const txt = locale.pages.viewRecord.version;
        const pid = 'UQ:1';
        const loadRecordToViewFn = jest.spyOn(ViewRecordActions, 'loadRecordVersionToView');
        useParams.mockImplementationOnce(() => ({ pid, version: recordVersionLegacy.rek_version }));
        const { getByTestId } = setup({
            viewRecordReducer: { recordToView: recordVersionLegacy, isDeletedVersion: true },
        });
        expect(loadRecordToViewFn).toHaveBeenCalledWith(pid, recordVersionLegacy.rek_version);
        expect(getByTestId(txt.alert.version.alertId)).toBeInTheDocument();
        expect(getByTestId(txt.alert.warning.alertId)).toBeInTheDocument();
    });

    it('should render deleted record correctly', () => {
        const { getByText } = setup({ viewRecordReducer: { isDeleted: true, recordToView: record } });
        expect(getByText('This work has been deleted.')).toBeInTheDocument();
        expect(
            getByText(
                'Long-range regulators of the lncRNA HOTAIR enhance its prognostic potential in breast cancer (default record)',
            ),
        ).toBeInTheDocument();
    });

    it('should render deleted data collection correctly', () => {
        const { getByText } = setup({
            viewRecordReducer: {
                isDeleted: true,
                recordToView: { ...record, rek_display_type: PUBLICATION_TYPE_DATA_COLLECTION },
            },
        });
        expect(getByText('This work has been deleted.')).toBeInTheDocument();
        expect(
            getByText(
                'Long-range regulators of the lncRNA HOTAIR enhance its prognostic potential in breast cancer (default record)',
            ),
        ).toBeInTheDocument();
    });

    it('should render deleted data collection with new doi correctly', () => {
        const newDoi = '10.000/abc';
        const { getByText, container } = setup({
            viewRecordReducer: {
                isDeleted: true,
                recordToView: {
                    ...record,
                    rek_display_type: PUBLICATION_TYPE_DATA_COLLECTION,
                    fez_record_search_key_new_doi: {
                        rek_new_doi: '10.000/abc',
                    },
                },
            },
        });
        expect(container.textContent).toContain(
            'This Data Collection has been deleted and substituted by another version.',
        );
        expect(container.querySelector(`[href="https://doi.org/${newDoi}"][target="_blank"]`)).toBeInTheDocument();
        expect(
            getByText(
                'Long-range regulators of the lncRNA HOTAIR enhance its prognostic potential in breast cancer (default record)',
            ),
        ).toBeInTheDocument();
    });

    it('should render deleted data collection with new doi and deletion notes correctly', () => {
        const newDoi = '10.000/abc';
        const deletionNotes = 'notes test';
        const { getByText, container } = setup({
            viewRecordReducer: {
                isDeleted: true,
                recordToView: {
                    ...record,
                    rek_display_type: PUBLICATION_TYPE_DATA_COLLECTION,
                    fez_record_search_key_new_doi: {
                        rek_new_doi: '10.000/abc',
                    },
                    fez_record_search_key_deletion_notes: {
                        rek_deletion_notes: deletionNotes,
                    },
                },
            },
        });
        expect(container.textContent).toContain(
            'This Data Collection has been deleted and substituted by another version.',
        );
        expect(container.querySelector(`[href="https://doi.org/${newDoi}"][target="_blank"]`)).toBeInTheDocument();
        expect(container.textContent).toContain(deletionNotes);
        expect(
            getByText(
                'Long-range regulators of the lncRNA HOTAIR enhance its prognostic potential in breast cancer (default record)',
            ),
        ).toBeInTheDocument();
    });

    it('should render deleted data collection with deletion notes correctly', () => {
        const deletionNotes = 'notes test';
        const { getByText, container } = setup({
            viewRecordReducer: {
                isDeleted: true,
                recordToView: {
                    ...record,
                    rek_display_type: PUBLICATION_TYPE_DATA_COLLECTION,
                    fez_record_search_key_deletion_notes: {
                        rek_deletion_notes: deletionNotes,
                    },
                },
            },
        });
        expect(getByText('This work has been deleted.')).toBeInTheDocument();
        expect(container.textContent).toContain(deletionNotes);
        expect(
            getByText(
                'Long-range regulators of the lncRNA HOTAIR enhance its prognostic potential in breast cancer (default record)',
            ),
        ).toBeInTheDocument();
    });

    it('should render loader', () => {
        const { getByText } = setup({ viewRecordReducer: { loadingRecordToView: true } });
        expect(getByText('Loading work')).toBeInTheDocument();
    });

    it('should render not found', () => {
        useParams.mockImplementationOnce(() => ({ pid: notFound }));
        const { container, queryByText } = setup();
        expect(container).toMatchSnapshot();
        expect(queryByText(locale.pages.workNotFound.title)).toBeInTheDocument();
        expect(queryByText(globalLocale.global.loginAlert.title)).not.toBeInTheDocument();
    });

    it('should render error', () => {
        const { getByText } = setup({
            viewRecordReducer: { recordToViewError: { message: 'PID not found', status: 403 } },
        });
        expect(getByText('You are not logged in -')).toBeInTheDocument();
        expect(getByText('Login to UQ eSpace for full search results and more services.')).toBeInTheDocument();
    });

    it('should render human readable message record not found', () => {
        const { getByText } = setup({
            viewRecordReducer: { recordToViewError: { message: 'PID not found', status: 404 } },
        });
        expect(getByText('Work not found')).toBeInTheDocument();
        expect(getByText('(404 - PID not found)')).toBeInTheDocument();
    });

    it('should have status prop in the header for admins', () => {
        window.matchMedia = createMatchMedia(window.innerWidth);
        userIsAdmin.mockImplementationOnce(() => true);
        const { getByText } = setup({
            viewRecordReducer: { recordToView: { ...record, rek_status: 1, rek_status_lookup: 'Unpublished' } },
        });
        expect(getByText('Unpublished')).toBeInTheDocument();
    });

    it('should load record to view', () => {
        const loadRecordToViewFn = jest.spyOn(ViewRecordActions, 'loadRecordToView');
        useParams.mockImplementationOnce(() => ({ pid: 'UQ:111111' }));
        setup({});
        expect(loadRecordToViewFn).toHaveBeenCalledWith('UQ:111111');
    });

    it('should reset store when component is unmounted', () => {
        const clearRecordToViewFn = jest.spyOn(ViewRecordActions, 'clearRecordToView');
        const { unmount } = setup({ viewRecordReducer: { recordToView: record } });
        unmount();
        expect(clearRecordToViewFn).toHaveBeenCalled();
    });

    it('should render NTRO Details', () => {
        const { getAllByText } = setup({
            viewRecordReducer: { recordToView: ntro },
            accountReducer: {
                account: {
                    canMasquerade: true,
                },
            },
        });
        expect(getAllByText('Scale/Significance of work').length).toBe(2);
    });

    it('should rerender component on props change', () => {
        const { getByText, queryByText, rerender } = setup({ viewRecordReducer: { loadingRecordToView: true } });
        expect(getByText('Loading work')).toBeInTheDocument();

        setup({ viewRecordReducer: { loadingRecordToView: false, recordToView: record } }, rerender);
        expect(queryByText('Loading work')).not.toBeInTheDocument();
    });

    it('redirects user to login if not Authorized', () => {
        const { location } = window;
        delete window.location;

        const assignFn = jest.fn();
        window.location = {
            assign: assignFn,
        };

        const { getByTestId } = setup({
            viewRecordReducer: { recordToViewError: { message: 'Your session has expired', status: 403 } },
        });

        fireEvent.click(getByTestId('action-button'));

        expect(assignFn).toHaveBeenCalledWith('https://fez-staging.library.uq.edu.au/login.php?url=dW5kZWZpbmVk');

        window.location = location;
    });

    it('should not render for researcher', () => {
        const { queryByTestId } = setup({
            viewRecordReducer: { recordToView: record },
            accountReducer: { account: accounts.uqresearcher },
        });
        expect(queryByTestId('adminViewRecordDrawerDesktop')).not.toBeInTheDocument();
        expect(queryByTestId('adminViewRecordDrawerMobile')).not.toBeInTheDocument();
        expect(queryByTestId('adminDrawerButton')).not.toBeInTheDocument();
    });
    it('should not render for student user', () => {
        const { queryByTestId } = setup({
            viewRecordReducer: { recordToView: record },
            accountReducer: { account: accounts.s1111111 },
        });
        expect(queryByTestId('adminViewRecordDrawerDesktop')).not.toBeInTheDocument();
        expect(queryByTestId('adminViewRecordDrawerMobile')).not.toBeInTheDocument();
        expect(queryByTestId('adminDrawerButton')).not.toBeInTheDocument();
    });

    describe('Admin record drawer for admins', () => {
        beforeEach(() => {
            userIsAdmin.mockImplementation(() => true);
        });

        it('should render for Admin user', () => {
            const { getByTestId } = setup({
                viewRecordReducer: { recordToView: record },
                accountReducer: { account: accounts.uqstaff },
            });
            expect(getByTestId('adminViewRecordDrawerDesktop')).toBeInTheDocument();
            expect(getByTestId('adminViewRecordDrawerMobile')).toBeInTheDocument();
            expect(getByTestId('btnAdminRecordDrawerCloseBtnDesktop')).not.toBeVisible();
            expect(getByTestId('btnAdminRecordDrawerCloseBtnMobile')).not.toBeVisible();
        });

        it('should open desktop admin drawer when button pressed', () => {
            const { getByTestId } = setup({
                viewRecordReducer: { recordToView: recordWithNotes },
                accountReducer: { account: accounts.uqstaff },
            });

            expect(getByTestId('adminViewRecordDrawerDesktop')).toBeInTheDocument();
            expect(getByTestId('btnAdminRecordDrawerCloseBtnDesktop')).not.toBeVisible();

            fireEvent.click(getByTestId('btnAdminToggleDrawerVisibility'));

            expect(getByTestId('btnAdminRecordDrawerCloseBtnDesktop')).toBeVisible();
            expect(getByTestId('btnAdminRecordDrawerCloseBtnMobile')).not.toBeVisible();
        });

        it('should open mobile admin drawer when button pressed', () => {
            window.matchMedia = createMatchMedia(320);

            const { getByTestId } = setup({
                viewRecordReducer: { recordToView: recordWithNotes },
                accountReducer: { account: accounts.uqstaff },
            });

            expect(getByTestId('adminViewRecordDrawerMobile')).toBeInTheDocument();
            expect(getByTestId('btnAdminRecordDrawerCloseBtnMobile')).not.toBeVisible();

            fireEvent.click(getByTestId('btnAdminToggleDrawerVisibility'));

            expect(getByTestId('btnAdminRecordDrawerCloseBtnDesktop')).not.toBeVisible();
            expect(getByTestId('btnAdminRecordDrawerCloseBtnMobile')).toBeVisible();
        });

        it('should render data in both admin drawers', () => {
            const { getByTestId } = setup({
                viewRecordReducer: { recordToView: recordWithAuthorAffiliates },
                accountReducer: { account: accounts.uqstaff },
            });

            expect(getByTestId('adminViewRecordDrawerDesktop')).toBeInTheDocument();
            expect(getByTestId('adminViewRecordDrawerMobile')).toBeInTheDocument();

            // DESKTOP
            // Notes
            expect(getByTestId('drawer-Desktop-content-scrollable-0-1')).toHaveTextContent(
                'Some internal notes for testing',
            );

            // Author affiliations
            expect(getByTestId('drawer-Desktop-content-value-2-1')).toHaveTextContent(
                'Valid author affiliation information has been added',
            );

            // WoS ID
            expect(getByTestId('drawer-Desktop-content-clipboard-4-1')).toHaveTextContent('000381303000009');
            // WoS DocType
            expect(getByTestId('drawer-Desktop-content-value-4-3')).toHaveTextContent('Article');

            // Scopus ID
            expect(getByTestId('drawer-Desktop-content-clipboard-6-1')).toHaveTextContent('2-s2.0-84975764277');
            // Scopus DocType
            expect(getByTestId('drawer-Desktop-content-value-6-3')).toHaveTextContent(
                'ar - Article (original research)',
            );

            // PubMed ID
            expect(getByTestId('drawer-Desktop-content-clipboard-8-1')).toHaveTextContent('27166757');
            // PubMed Central ID
            expect(getByTestId('drawer-Desktop-content-clipboard-8-3')).toHaveTextContent('PMC5179926');
            // PubMed DocType
            expect(getByTestId('drawer-Desktop-content-value-8-5')).toHaveTextContent(
                'Journal Article - Article (original research)',
            );

            // MOBILE
            // Notes
            expect(getByTestId('drawer-Mobile-content-scrollable-0-1')).toHaveTextContent(
                'Some internal notes for testing',
            );

            // Author affiliations
            expect(getByTestId('drawer-Mobile-content-value-2-1')).toHaveTextContent(
                'Valid author affiliation information has been added',
            );

            // WoS ID
            expect(getByTestId('drawer-Mobile-content-clipboard-4-1')).toHaveTextContent('000381303000009');
            // WoS DocType
            expect(getByTestId('drawer-Mobile-content-value-4-3')).toHaveTextContent('Article');

            // Scopus ID
            expect(getByTestId('drawer-Mobile-content-clipboard-6-1')).toHaveTextContent('2-s2.0-84975764277');
            // Scopus DocType
            expect(getByTestId('drawer-Mobile-content-value-6-3')).toHaveTextContent(
                'ar - Article (original research)',
            );

            // PubMed ID
            expect(getByTestId('drawer-Mobile-content-clipboard-8-1')).toHaveTextContent('27166757');
            // PubMed Central ID
            expect(getByTestId('drawer-Mobile-content-clipboard-8-3')).toHaveTextContent('PMC5179926');
            // PubMed DocType
            expect(getByTestId('drawer-Mobile-content-value-8-5')).toHaveTextContent(
                'Journal Article - Article (original research)',
            );
        });
    });

    describe('datastream visibility', () => {
        const filename = 'image.jpg';
        const publication = {
            ...record,
            fez_datastream_info: [
                {
                    dsi_pid: 'UQ:792099',
                    dsi_dsid: filename,
                    dsi_embargo_date: null,
                    dsi_open_access: null,
                    dsi_label: 'testing image description',
                    dsi_mimetype: 'image/jpeg',
                    dsi_copyright: null,
                    dsi_state: 'A',
                    dsi_size: 97786,
                    dsi_security_policy: 1,
                    dsi_security_inherited: 0,
                },
            ],
        };
        publication.fez_record_search_key_author_id[0].rek_author_id = currentAuthor.uqresearcher.data.aut_id;

        it("shouldn't show admin only attachments to authors", () => {
            const { queryByText } = setup({
                viewRecordReducer: { recordToView: publication },
                accountReducer: { account: accounts.uqresearcher, author: currentAuthor.uqresearcher.data },
            });
            expect(queryByText(filename)).not.toBeInTheDocument();
        });

        it('should show admin only attachments to admin', () => {
            userIsAdmin.mockImplementation(() => true);
            const { queryByText } = setup({
                viewRecordReducer: { recordToView: publication },
                accountReducer: { account: accounts.uqresearcher, author: currentAuthor.uqresearcher.data },
            });
            expect(queryByText(filename)).toBeInTheDocument();
        });

        describe('NTRO', () => {
            const ntroPublication = {
                ...ntro,
                fez_record_search_key_author_id: publication.fez_record_search_key_author_id,
                fez_datastream_info: publication.fez_datastream_info,
            };

            it('should show admin only attachments to authors of NTROs (non research reports)', () => {
                const filename = 'image.jpg';
                const { queryByText } = setup({
                    viewRecordReducer: { recordToView: ntroPublication },
                    accountReducer: { account: accounts.uqresearcher, author: currentAuthor.uqresearcher.data },
                });
                expect(queryByText(filename)).toBeInTheDocument();
            });

            it("shouldn't show admin only attachments to authors of NTROs research reports", () => {
                const { queryByText } = setup({
                    viewRecordReducer: {
                        recordToView: { ...ntroPublication, rek_subtype: NTRO_SUBTYPE_RREB_PUBLIC_SECTOR },
                    },
                    accountReducer: { account: accounts.uqresearcher, author: currentAuthor.uqresearcher.data },
                });
                expect(queryByText(filename)).not.toBeInTheDocument();
            });
        });
    });
});
