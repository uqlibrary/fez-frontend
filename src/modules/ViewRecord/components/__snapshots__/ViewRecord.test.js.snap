// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewRecord should render not found 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container StandardPage css-11lq3yg-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
    >
      <h2
        class="MuiTypography-root MuiTypography-body1 css-v2c8i0-MuiTypography-root"
        data-analyticsid="page-title"
        data-testid="page-title"
        id="page-title"
      >
        Work not found
      </h2>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    />
    <div
      class="MuiGrid-root MuiGrid-item css-1guqyqn-MuiGrid-root"
    >
      <div
        class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row css-vdigti-MuiGrid2-root"
        data-testid="notFoundGridContainer"
        id="notFoundGridContainer"
      >
        <div
          class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 css-16fnxy3-MuiGrid2-root"
        >
          <div>
            <p>
              The work you are attempting to access does not appear in our system.
            </p>
            <p>
              If you believe this is in error, please
               
              <a
                href="https://guides.library.uq.edu.au/for-researchers/uqespace-publications-datasets/contact-us"
              >
                contact the eSpace team.
              </a>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
